test/attn.py

```diff
@@ -0,0 +1,70 @@
+from torch import autograd
+import torch
+
+# Inherit from Function
+class AttentionFunction(autograd.Function):
+    @staticmethod
+    def attn(ctx, q, k, v):
+        a = torch.tanh(torch.matmul(q, k.T))
+        ctx.save_for_backward(q, k, v, a)
+        o = torch.matmul(a, v)
+        return o, a
+
+    # Note that both forward and backward are @staticmethods
+    @staticmethod
+    # bias is an optional argument
+    def forward(ctx, q, k, v):
+        assert len(q.size()) == 2, f"input q must be 2D but instead has {len(q.size())} dims"
+        assert len(k.size()) == 2, f"input k must be 2D but instead has {len(k.size())} dims"
+        assert len(v.size()) == 2, f"input v must be 2D but instead has {len(v.size())} dims"
+        assert q.size(0) == k.size(0) and k.size(0) == v.size(0), "all inputs must have the same first dimension"
+        assert q.size(1) == k.size(1), "q and k must share the same size for their second dimension"
+        ctx.set_materialize_grads(False)
+        return AttentionFunction.attn(ctx, q, k, v)
+
+    # This function has TWO outputs, so it gets TWO gradients :D
+    @staticmethod
+    def backward(ctx, grad_o, grad_a):
+        # This is a pattern that is very convenient - at the top of backward
+        # unpack saved_tensors and initialize all gradients w.r.t. inputs to
+        # None. Thanks to the fact that additional trailing Nones are
+        # ignored, the return statement is simple even when the function has
+        # optional inputs.
+        q, k, v, a = ctx.saved_tensors
+        total_q = total_k = total_v = None
+
+        # These needs_input_grad checks are optional and there only to
+        # improve efficiency. If you want to make your code simpler, you can
+        # skip them. Returning gradients for inputs that don't require it is
+        # not an error.
+
+        i = grad_o @ v.T if grad_o is not None else torch.zeros(1)
+        j = 1 - a ** 2
+        if ctx.needs_input_grad[0]:
+            total_q = (i * j) @ k + ((grad_a * j) @ k if grad_a is not None else torch.zeros(1))
```

Comment:
A few more small comments (no action needed!):
- FYI another way we represent zeros with a special object called ZeroTensor which reduces unnecessary computation when doing certain operations like multiplying or adding See https://github.com/pytorch/pytorch/issues/69687 for more context (currently it is mostly used for forward AD)
- For this case specifically I would just conditionally add the second term `(grad_a * j) @ k` if it is defined to avoid materializing that extra zero
- naming nit: It might be better to use upper case letters like `A`, `V`, `Q`, to refer to 2-D tensors (matrices), vs i, j which tends to refer to indices

Reply:
acknowledged! zerotensors seem cool